fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneConcurrency.java

```diff
@@ -0,0 +1,162 @@
+     * This implementation does NOT perform exception mapping (as the original {@link FDBRecordContext#asyncToSync} does).
+     * This method is meant to be used internally in places where obtaining and using the result of asynchronous
+     * operation is required.
+     * This method uses the {@link FDBDatabase#getAsyncToSyncTimeout} to find the period to use for the timeout.
+     * This method will throw the runtime exception that was thrown by the Future's realization in case such error
+     * occurred.
+     * This method will use the legacy  {@link FDBRecordContext#asyncToSync} if the LUCENE_USE_LEGACY_ASYNC_TO_SYNC
+     * property is set to TRUE (the default)
+     *
+     * @param event the timer event to use for recording the waits
+     * @param async the future to wait on
+     * @param recordContext the context to use for callback, recording the event and getting the timeout
+     *
+     * @return the result of the future's operation
+     */
+    @Nullable
+    @API(API.Status.INTERNAL)
+    public static <T> T asyncToSync(@Nonnull StoreTimer.Wait event, @Nonnull CompletableFuture<T> async, @Nonnull FDBRecordContext recordContext) {
+        if (recordContext.getPropertyStorage().getPropertyValue(LUCENE_USE_LEGACY_ASYNC_TO_SYNC)) {
+            return recordContext.asyncToSync(event, async);
+        }
+
+        if (recordContext.hasHookForAsyncToSync() && !MoreAsyncUtil.isCompletedNormally(async)) {
+            recordContext.getHookForAsyncToSync().accept(event);
+        }
+
+        recordContext.getDatabase().checkIfBlockingInFuture(async);
+
+        if (async.isDone()) {
+            try {
+                return async.get();
+            } catch (ExecutionException ex) {
+                throw FDBExceptions.wrapException(ex);
+            } catch (InterruptedException ex) {
+                Thread.currentThread().interrupt();
+                throw FDBExceptions.wrapException(ex);
+            }
+        } else {
+            final Pair<Long, TimeUnit> asyncToSyncTimeout = recordContext.getDatabase().getAsyncToSyncTimeout(event);
+            Timeout timeout = (asyncToSyncTimeout == null) ? null : Timeout.of(asyncToSyncTimeout.getLeft(), asyncToSyncTimeout.getRight());
+            final FDBStoreTimer timer = recordContext.getTimer();
+            final long startTime = System.nanoTime();
+            try {
+                if (timeout != null) {
+                    return async.get(timeout.getDuration(), timeout.getTimeUnit());
+                } else {
+                    return async.get();
+                }
+            } catch (TimeoutException ex) {
+                if (timer != null) {
+                    timer.recordTimeout(event, startTime);
+                    throw new AsyncToSyncTimeoutException(ex.getMessage(), ex, LogMessageKeys.TIME_LIMIT.toString(), timeout.getDuration(), LogMessageKeys.TIME_UNIT.toString(), timeout.getTimeUnit());
```

Comment:
Nit: For readability to ensure that there is a correct set of pairs, it would be good to wrap this line
```suggestion
                    throw new AsyncToSyncTimeoutException(ex.getMessage(), ex,
                            LogMessageKeys.TIME_LIMIT.toString(), timeout.getDuration(),
                            LogMessageKeys.TIME_UNIT.toString(), timeout.getTimeUnit());
```

Reply:
Done.